crates/signs/src/line.rs

```diff
@@ -0,0 +1,209 @@
+    SpawnLines,
+}
+
+fn setup(mut commands: Commands) {
+    commands.init_resource::<LineEntities>();
+    commands.init_resource::<LineTransforms>();
+}
+
+fn cleanup(mut commands: Commands) {
+    commands.remove_resource::<LineEntities>();
+    commands.remove_resource::<LineTransforms>();
+}
+
+pub struct UpdateLineVisibilityEvent {
+    owner: Entity,
+    visible: bool,
+}
+
+impl UpdateLineVisibilityEvent {
+    pub fn new(owner: Entity, visible: bool) -> Self {
+        Self { owner, visible }
+    }
+}
+
+pub struct UpdateLineLocationEvent {
+    owner: Entity,
+    start: Vec3,
+    end: Vec3,
+}
+
+impl UpdateLineLocationEvent {
+    pub fn new(owner: Entity, start: Vec3, end: Vec3) -> Self {
+        Self { owner, start, end }
+    }
+}
+
+pub struct SpawnLineEvent(Transform, Entity);
+
+#[derive(Resource, Default)]
+struct LineEntities(AHashMap<Entity, Entity>);
+
+#[derive(Resource, Default)]
+struct LineTransforms(AHashMap<Entity, Transform>);
+
+fn update_line_location(
+    lines: Res<LineEntities>,
+    mut events: EventReader<UpdateLineLocationEvent>,
+    mut transforms: Query<&mut Transform>,
+    mut line_transforms: ResMut<LineTransforms>,
+    mut spawn_line_events: EventWriter<SpawnLineEvent>,
+) {
+    for event in &mut events {
+        let matrix = compute_line_transform(event.start, event.end);
+        let transform = Transform::from_matrix(matrix);
+        line_transforms.0.insert(event.owner, transform);
+        if let Some(line_entity) = lines.0.get(&event.owner) {
+            *transforms
+                .get_mut(*line_entity)
+                .expect("Line should have a transform") = transform
+        } else {
+            spawn_line_events.send(SpawnLineEvent(transform, event.owner))
+        }
+    }
+}
+
+fn update_line_visibility(
+    mut commands: Commands,
+    mut events: EventReader<UpdateLineVisibilityEvent>,
+    mut lines: ResMut<LineEntities>,
+    line_transforms: Res<LineTransforms>,
+    mut spawn_line_events: EventWriter<SpawnLineEvent>,
+) {
+    for event in &mut events {
+        let line_entity = lines.0.entry(event.owner);
+        if event.visible && matches!(line_entity, Entry::Vacant(_)) {
+            if let Some(transform) = line_transforms.0.get(&event.owner) {
+                spawn_line_events.send(SpawnLineEvent(*transform, event.owner));
+            }
+        } else if !event.visible {
+            if let Entry::Occupied(entry) = line_entity {
+                commands.entity(entry.remove()).despawn_recursive();
+            }
+        }
+    }
+}
+
+fn spawn_line(
+    mut lines: ResMut<LineEntities>,
+    mut events: EventReader<SpawnLineEvent>,
+    mut commands: Commands,
+    mut meshes: ResMut<Assets<Mesh>>,
+    mut materials: ResMut<Assets<LineMaterial>>,
+) {
+    for SpawnLineEvent(transform, owner) in &mut events {
+        let line_id = commands
+            .spawn(MaterialMeshBundle {
+                mesh: meshes.add(shape::Plane::from_size(1.0).into()),
```

Comment:
Performance-wise, it would be better to add the mesh only once (during `setup()`) and reuse it. That way, it would be stored only once in the VRAM and instancing would be used.